### PR TITLE
refactor(api-markdown-documenter): Update HTML rendering functionality to leverage HAST domain and existing HAST rendering libraries

### PR DESCRIPTION
**Pull request by Josmithr**
Removes A LOT of bespoke HTML rendering logic, replacing it with [hast-util-to-html](https://www.npmjs.com/package/hast-util-to-html) and [hast-util-format](https://www.npmjs.com/package/hast-util-format) to do the rendering on our behalf.

Note that there are A LOT of test asset diffs here. The rendering libraries mentioned above make different styling choices than we made in our bespoke rendering. In most cases these changes are trivial, or arguably readability improvements. Regardless, this PR makes the formatting no longer a direct responsibility of this library, which is a win IMO.

### CODE REVIEW

File: tools/api-markdown-documenter/README.md
Comment (Josmithr) [inline tools/api-markdown-documenter/README.md:164]: I really like how the HAST domain intermediary has worked out (and it aligns with the original intentions for this library). I'm hoping to get the MDAST side of things in the near future (probably on my own time, because I'm a nerd).

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/index.html
Comment (Josmithr) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/index.html:1]: For reviewers: most of the formatting changes boil down to:

- Casing of `doctype`
- Not line wrapping lowest level tag blocks (mostly paragraphs, headings, etc.)

File: tools/api-markdown-documenter/README.md
Comment (alexvy86) [inline tools/api-markdown-documenter/README.md:147]:
```suggestion
End to end, this library can be viewed as a pseudo-functional transformation pipeline mapping from an [APIModel][] generated by `API-Extractor` to one or more `Markdown` or `HTML` formatted document files.
```

Also, curious, does the empty "link target" for APIModel serve a purpose?

File: tools/api-markdown-documenter/src/renderers/markdown-renderer/default-renderers/RenderHeading.ts
Comment (alexvy86) [inline tools/api-markdown-documenter/src/renderers/markdown-renderer/default-renderers/RenderHeading.ts:86]: I was thinking that in an ideal world I'd expect to see anchors rendered in markdown with the usual syntax... which I then realized actually means "headings (and only headings) are anchors". But I guess anchors don't _have_ to be tied to headings, and this gives us that flexibility?

File: tools/api-markdown-documenter/src/renderers/markdown-renderer/test/RenderSpan.test.ts
Comment (alexvy86) [inline tools/api-markdown-documenter/src/renderers/markdown-renderer/test/RenderSpan.test.ts:73]: Bug fix, it seems?

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/index.html
Comment (alexvy86) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/index.html:21]: Is this correct? Seems to have mangled the `<` for the `<b>` elements.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html
Comment (alexvy86) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html:10]: Here a `&gt;` changed to a `>` which feels like it'd break something when rendering this HTML doc.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testenum-enum.html
Comment (alexvy86) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testenum-enum.html:30]: Not sure I love single-lines for the `<code>` elements. I'm also not sure if those respect formatting like `<pre>` would, so what would happen with multi-line contents (for both). The files I've looked at so far don't have an example of that, might be good to add.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testenum-testenumvalue1-enummember.html
Comment (alexvy86) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testenum-testenumvalue1-enummember.html:13]: I think going from `&quot;` to `"` is fine as long as we're in the contents of an HTML element... so this has no impact? But might be worth double checking.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninginlinetype-function.html
Comment (alexvy86) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninginlinetype-function.html:13]: Oh, multi-line `<code>` requires `<br>` then? Still, if we can configure to force those to not be single-line I think it'd produce more readable HTML. Won't block on it if we can't, though.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturningintersectiontype-function.html
Comment (alexvy86) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturningintersectiontype-function.html:13]: Here the second paragraph got italicized as a whole, whereas only the non-`<a>` text was before. I think the new version is fine... but in case you want to look into it.

Not a fan of multiple `<p>`s in a single line, but another thing I wouldn't block on; I'd just be happier if we can configure the behavior to not do it. And I see in lines 20-21 below that it can render them in separate lines, so this feels weird.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturningintersectiontype-function.html
Comment (alexvy86) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturningintersectiontype-function.html:21]: Mangled `&lt;` and (incorrectly, I think) transformed `&gt;` into `>`.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninguniontype-function.html
Comment (alexvy86) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninguniontype-function.html:18]: `&#124; -> |`. Not sure that `|` has any special meaning in HTML syntax so I think it might be ok, just like for the quotes in a previous comment, but also worth checking, since some of the changes of this kind do seem incorrect to me.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testnamespace-testclass-testclassmethod-method.html
Comment (alexvy86) [inline tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testnamespace-testclass-testclassmethod-method.html:37]: Again `&lt;` and `&gt;` mangling.

**Review by alexvy86 — COMMENTED** [review]
Leaving a few comments which I think might result in changes, before I look through the rest of the files.

File: tools/api-markdown-documenter/README.md
Comment (Josmithr) [reply tools/api-markdown-documenter/README.md:147]: That's a supported syntax in GitHub flavored markdown. It corresponds to a definition at the bottom of the file. Lets you reuse link targets without having to specify them everywhere.

File: tools/api-markdown-documenter/src/renderers/markdown-renderer/default-renderers/RenderHeading.ts
Comment (Josmithr) [reply tools/api-markdown-documenter/src/renderers/markdown-renderer/default-renderers/RenderHeading.ts:86]: Exactly. Both HTML and Markdown only support heading levels 1-6. Our system tries to offer more flexibility than that by rendering higher level "headings" as bold text with anchors.

Note that this is existing behavior - just moved the function to be more local now that it's only used in 1 place.

File: tools/api-markdown-documenter/src/renderers/markdown-renderer/test/RenderSpan.test.ts
Comment (Josmithr) [reply tools/api-markdown-documenter/src/renderers/markdown-renderer/test/RenderSpan.test.ts:73]: No, just a formatting difference. / making the "expected" string a bit more consistent by not leveraging an existing variable in one place but not the other.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/index.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/index.html:21]: Oh, good catch. Yeah, this is incorrect. The code is incorrectly escaping HTML that came from the code comment itself (which is supposed to be supported). I'll add a unit test for this case as well. This is important.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html:10]: I believe the spaces around `>` make it safe in HTML. I'll confirm this, though. I believe only `<` characters need to be unilaterally escaped, and `>` is more context dependent. But I'll double check this.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninginlinetype-function.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninginlinetype-function.html:13]: Not sure `hast-util-format` lets you customize to that degree, but I'll check.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/index.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/index.html:21]: Okay, so after digging into this for a while, I have formed an opinion on something I think I will change (in a separate PR). The logic that was broken here was our use of unsafe HTML. TSDoc parses HTML tags and passes them to us as "escaped text", with the expectation that they will be forwarded to the renderer untouched. This is somewhat sketchy to begin with.

Secondly, the tree structure they give is bananas. For the comment text `<b>Hello world!</b>`, the tree TSDoc generates is:
```json
[
    { type: "HtmlStartTag", "tag": "b" }
    { type: "EscapedText", text: "Hello world!" },
    { type: "HtmlCloseTag", "tag": "b"}
]
```

rather than something like:

```json
{
    "type": "HtmlBlock",
    "tag": "b",
    "children": [
        { "type": "text", "text": "Hello world!" }
    ]
}
```

This makes it difficult to preserve intention when mapping to our output domains. We got around this by treating all of the adjacent nodes as "escaped" and letting the contents get mapped to output raw. Sketchy, but it's what the base `api-documenter` library does.

I double checked, and intellisense simply ignores HTML tags in comments. It renders the inner contents, but completely ignores any tags decorating them.

My opinion: do this. It's safer, better aligns with TSDoc spec (which makes no claims about support for embedded HTML AFAIK), and makes our transformation logic A LOT simpler.

I'll put up a separate PR that makes this change.

Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/index.html:21]: PR: https://github.com/microsoft/FluidFramework/pull/22671

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninginlinetype-function.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninginlinetype-function.html:13]: No, it doesn't. But I still prefer to go this route over managing our own rendering. If a user doesn't like this formatting, they can choose their own `hast` renderer to use.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturningintersectiontype-function.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturningintersectiontype-function.html:13]: > Here the second paragraph got italicized as a whole, whereas only the non-<a> text was before. I think the new version is fine... but in case you want to look into it.

This appears to have been a bug in our previous rendering code. The link _should_ have respected the formatting of the containing span.

> Not a fan of multiple <p>s in a single line, but another thing I wouldn't block on; I'd just be happier if we can configure the behavior to not do it. And I see in lines 20-21 below that it can render them in separate lines, so this feels weird.

I agree. But I'm viewing this renderer as a default implementation that users can take or leave. So I'm not sure I personally care too much what the output looks like. Users can leverage the `hast` output and call / implement their own renderers as they see fit.

For fluidframework.com, we only leverage this functionality for tables, and I was intending to likely not opt-in to formatted rendering long term anyways for performance reasons.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testnamespace-testclass-testclassmethod-method.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testnamespace-testclass-testclassmethod-method.html:37]: Confirmed that this works correctly via ff.com build.
image

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturningintersectiontype-function.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturningintersectiontype-function.html:21]: This escaping appears to be valid.

image

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninguniontype-function.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninguniontype-function.html:18]: Yup, this is valid.
image

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testenum-testenumvalue1-enummember.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testenum-testenumvalue1-enummember.html:13]: Yeah, `"` is valid when it isn't within the context of a tag declaration. Confirmed that this is rendering correctly on the website (with these changes applied).

image

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testenum-enum.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testenum-enum.html:30]: The code blocks would include `<br>` tags for line breaks, currently. Good call on adding a multi-line code sample to the test assets, though. I'll put up a separate PR that does that.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html:10]: Confirmed.

image

File: tools/api-markdown-documenter/src/renderers/markdown-renderer/test/RenderSpan.test.ts
Comment (Josmithr) [reply tools/api-markdown-documenter/src/renderers/markdown-renderer/test/RenderSpan.test.ts]: This change was actually reverted with my most recent changes. I realized that the way I had implemented styling was incorrect. Many of these such differences are now gone.

**Comment by Josmithr** [comment]
Just put up a separate PR that fixes an issue I detected with how we apply text formatting attributes: #22672. I'll wait to proceed further with this PR until that has been merged.

File: tools/api-markdown-documenter/README.md
Comment (alexvy86) [reply tools/api-markdown-documenter/README.md]: Oh, I know those, but didn't know you could make the text itself the "id". I see now 👍.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html
Comment (alexvy86) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html:10]: The confirmation is on the brackets for the generic type? The opening one is `&lt;` and the closing one `>`?

If so, good to know it works... I think I'd still prefer them both escaped for consistency if possible, but won't block on it.

File: tools/api-markdown-documenter/src/renderers/html-renderer/Render.ts
Comment (alexvy86) [inline tools/api-markdown-documenter/src/renderers/html-renderer/Render.ts:65]:
```suggestion
		// Note: this API is specifically typed to only accept a `Root` node, but its code only requires any `Nodes`.
```

**Review by alexvy86 — APPROVED** [review]
Looks good!

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/default-config/test-suite-a/testemptyinterface-interface.html:10]: That's correct. And yeah, I agree with you there. Unfortunately, `hast-util-to-html` doesn't let you configure this, and they seem to have taken the policy of minimal escaping. Overall, I prefer leveraging the external tool over maintaining our own rendering. And it's a widely used library, so we can have some confidence that they're on top of issues.

File: tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninginlinetype-function.html
Comment (Josmithr) [reply tools/api-markdown-documenter/src/test/snapshots/html/simple-suite-test/sparse-config/test-suite-a/testfunctionreturninginlinetype-function.html:13]: Test collateral has been updated. Confirmed that mutli-line code blocks preserve the necessary `<br>` tags.

**Pull request merged**
